sockeye/scoring.py

```diff
@@ -0,0 +1,474 @@
+                                                                   data_iter=data_iter,
+                                                                   mapid=mapid)
+            scored_samples.append(single_model_scored_samples)
+
+        return self._make_outputs(scored_samples)
+
+    def _make_outputs(self,
+                      scored_samples: ScoredSamples) -> List[ScoringOutput]:
+        """
+        Generates a list of ScoringOutputs from scorer output.
+
+        :param scored_samples: A list of scored batches.
+        :return: A list of ScoringOutputs that can be handled by an output
+        handler.
+        """
+        scored_outputs = []
+
+        for sample in zip(*scored_samples):
+            scores = [score for (sentence_id, score) in sample]
+            ensemble_score = self._linear_interpolation(scores)
+
+            sentence_id = sample[0][0]
+
+            scored_output = ScoringOutput(sentence_id=sentence_id,
+                                          source_tokens=[],
+                                          target_tokens=[],
+                                          score=ensemble_score)
+            scored_outputs.append(scored_output)
+
+        return scored_outputs
+
+
+def load_models(context: mx.context.Context,
+                batch_size: int,
+                model_folders: List[str],
+                data_iters: List[data_io.BaseParallelSampleIter],
+                configs: List[data_io.DataConfig],
+                checkpoints: Optional[List[int]] = None,
+                bucketing: bool = False) -> List[ScoringModel]:
+
+    """
+    Loads a list of models for scoring.
+
+    :param context: MXNet context to bind modules to.
+    :param batch_size: Batch size.
+    :param model_folders: List of model folders to load models from.
+    :param data_iter: List of iterators over the data set (one per model, can use different vocabularies).
+    :param configs: List of configuration objects holding details about the models (one per model).
+    :param checkpoints: List of checkpoints to use for each model in model_folders. Use None to load best checkpoint.
+    :param bucketing: If True bucketing will be used, if False the computation
+    graph will always be unrolled to the full length.
+    :return: List of models.
+    """
+
+    models = []
+    if checkpoints is None:
+        checkpoints = [None] * len(model_folders)
+
+    for model_folder, checkpoint, config, data_iter in zip(model_folders, checkpoints, configs, data_iters):
+
+        model = ScoringModel(model_folder=model_folder,
+                             context=context,
+                             data_iter=data_iter,
+                             config=config,
+                             checkpoint=checkpoint,
+                             bucketing=bucketing)
+        models.append(model)
+
+    return models
+
+
+def create_data_iter_and_vocab(args: argparse.Namespace,
```

Comment:
So scoring in general has things in common with both training (a single MXNet Module capturing the entire computation graph) and inference (translation one sentence at a time and making sure to have the correct order on the output).

In the code below you make use of a lot of the training data loading code. I'm wondering if it would be easier to factor out some of the logic in `translate.py` and reuse it here instead. For example you could make use of `translate.py.Translator._get_inference_input`, so that you could simply have a for loop over sentences and then from each sentence (or list of sentences) construct an NDArray. This would get around using RawParallelDatasetLoader and putting things in the correct order in the end.

What do you think?

Reply:
Thanks for submitting this! I think it would be very useful and would like to have this merged in. However, looking over the code, it seems much more extensive than necessary. Why not just augment `translate.py` and then force the choice of the word at each point? To sketch this out:

- Require the JSON format, using something like:

      { 'text': "This is a test', 'force': "This is its translation" }

   You then pass `--json-input` and it will parse this out, making a `force` field (ha) available

- In `inference.py`, you'd add a field to `TranslatorInput` and turn the `force` field (still funny) to IDs

- Then, instead of calling `_beam_search()`, you could call something like `_force_decode()`, which would copy some code from `_beam_search()` (like the loop over `t` and the call to `_decode_step()`) to get the probs, but would then choose the appropriate column instead of calling `topk()`

That would eliminate a lot of extra files and redundancy.

Reply:
Hi Tobi and Matt,
Thanks for your comments - I did not forget about this PR;
I agree that  much of the inference code could be reused, hopefully I will soon find time to implement your suggestions.

Reply:
In general I don't mind having a separate module for scoring. Scoring has a lot more in common with training than it was with inference, as like in training you can create the complete computation graph. Force decoding in `inference.py` adds a lot of unnecessary overhead. At the same time we could try to share the logic for creating batches from inference. So to some extend it makes sense to share code from both training (computation graph construction) and inference (data_io). 
So in general I'd vote for keeping the separate scoring module, but maybe trying to factor out the needed functionalities from the training and inference modules.

Reply:
Just a wild shot here without any specific pointers at code, but when PPL is computed during training this in fact force decoding. Could we take advantage of it?

Reply:
the main reason to re-use the data iterator from training was to not have to deal with bucketing for source-target pairs - the only difference to training is that we need to keep track of the original input order. scoring really has not much in common with inference, except the ordering, and a 'forced' decoding seems overly complicated when you can just use decode_sequence?

Reply:
I agree that conceptually it's closer to training since the source and target are fixed. But I think this could be done in about 20 lines of code as an inference task (where bucketing and batching will come along for free), and I think it would take a lot more as an inference task, and that appeals to me.